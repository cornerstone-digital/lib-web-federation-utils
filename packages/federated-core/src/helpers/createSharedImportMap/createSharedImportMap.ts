import { SharedImportMapConfig } from './createSharedImportMap.types'

const getDepMap = (format: 'umd' | 'esm') => ({
  vue: {
    dev: `vue/3.2.25/vue.global.prod.min.js`,
    prod: `vue/3.2.25/vue.global.prod.min.js`,
  },
  react: {
    dev: `react/17.0.2/${format}/react.production.min.js`,
    prod: `react/17.0.2/${format}/react.production.min.js`,
  },

  reactDom: {
    dev: `react-dom/17.0.2/${format}/react-dom.production.min.js`,
    prod: `react-dom/17.0.2/${format}/react-dom.production.min.js`,
  },

  reactIs: {
    dev: `react-is/17.0.2/${format}/react-is.production.min.js`,
    prod: `react-is/17.0.2/${format}/react-is.production.min.js`,
  },

  styledComponents: {
    dev: `styled-components/5.3.5/styled-components.min.js`,
    prod: `styled-components/5.3.5/styled-components.min.js`,
  },
})

export const getImportUrl = (url: string, baseUrl: string) => {
  if (url.startsWith('http')) {
    return url
  }

  return `${baseUrl}/${url}`
}

const createSharedImportMap = (
  config: SharedImportMapConfig,
  imports: Record<string, string> = {}
) => {
  const depMap = getDepMap(config.format)

  if (config.imports.addVue) {
    imports['vue'] = getImportUrl(
      config.isDev ? depMap.vue.dev : depMap.vue.prod,
      config.basePath
    )
  }
  if (config.imports.addReact) {
    imports['react'] = getImportUrl(
      config.isDev ? depMap.react.dev : depMap.react.prod,
      config.basePath
    )
    imports['react-dom'] = getImportUrl(
      config.isDev ? depMap.reactDom.dev : depMap.reactDom.prod,
      config.basePath
    )
    imports['react-is'] = getImportUrl(
      config.isDev ? depMap.reactIs.dev : depMap.reactIs.prod,
      config.basePath
    )
  }

  if (config.imports.addStyledComponents) {
    imports['styled-components'] = getImportUrl(
      config.isDev ? depMap.styledComponents.dev : depMap.styledComponents.prod,
      config.basePath
    )
  }

  return imports
}

export default createSharedImportMap
