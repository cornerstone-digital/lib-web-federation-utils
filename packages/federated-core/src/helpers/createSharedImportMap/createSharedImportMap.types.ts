export type SharedImportMapConfig = {
  imports: {
    addVue: boolean
    addReact: boolean
    addStyledComponents: boolean
  }
  isDev: boolean
  basePath: string
  format: 'umd' | 'esm'
}
