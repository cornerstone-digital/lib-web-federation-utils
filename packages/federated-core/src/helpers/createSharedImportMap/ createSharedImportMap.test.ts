import createSharedImportMap, { getImportUrl } from './createSharedImportMap'

describe('createSharedImportMap', () => {
  it('should return import map with prod when isDev false', () => {
    const importMap = createSharedImportMap({
      imports: {
        addVue: true,
        addReact: true,
        addStyledComponents: true,
      },
      format: 'umd',
      isDev: false,
      basePath: 'http://cdn.vodadone.co.uk/shared-assets',
    })

    expect(importMap).toMatchSnapshot()
  })

  it('should return import map with dev when isDev true', () => {
    const importMap = createSharedImportMap({
      imports: {
        addVue: true,
        addReact: true,
        addStyledComponents: true,
      },
      format: 'umd',
      isDev: true,
      basePath: 'http://cdn.vodadone.co.uk/shared-assets',
    })

    expect(importMap).toMatchSnapshot()
  })

  describe('getImportUrl', () => {
    it('should return url with basePath', () => {
      const url = getImportUrl(
        'react/17.0.2/umd/react.production.min.js',
        'http://cdn.vodadone.co.uk/shared-assets'
      )
      expect(url).toBe(
        'http://cdn.vodadone.co.uk/shared-assets/react/17.0.2/umd/react.production.min.js'
      )
    })

    it('should not add basePath when url starts with http', () => {
      const url = getImportUrl(
        'http://some-other-url/react/17.0.2/umd/react.production.min.js',
        'http://cdn.vodadone.co.uk/shared-assets'
      )
      expect(url).toBe(
        'http://some-other-url/react/17.0.2/umd/react.production.min.js'
      )
    })
  })
})
