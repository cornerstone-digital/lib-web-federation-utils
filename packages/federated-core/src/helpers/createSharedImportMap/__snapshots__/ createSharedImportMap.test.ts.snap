// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSharedImportMap should return import map with dev when isDev true 1`] = `
Object {
  "react": "http://cdn.vodadone.co.uk/shared-assets/react/17.0.2/umd/react.production.min.js",
  "react-dom": "http://cdn.vodadone.co.uk/shared-assets/react-dom/17.0.2/umd/react-dom.production.min.js",
  "react-is": "http://cdn.vodadone.co.uk/shared-assets/react-is/17.0.2/umd/react-is.production.min.js",
  "styled-components": "http://cdn.vodadone.co.uk/shared-assets/styled-components/5.3.5/styled-components.min.js",
  "vue": "http://cdn.vodadone.co.uk/shared-assets/vue/3.2.25/vue.global.prod.min.js",
}
`;

exports[`createSharedImportMap should return import map with prod when isDev false 1`] = `
Object {
  "react": "http://cdn.vodadone.co.uk/shared-assets/react/17.0.2/umd/react.production.min.js",
  "react-dom": "http://cdn.vodadone.co.uk/shared-assets/react-dom/17.0.2/umd/react-dom.production.min.js",
  "react-is": "http://cdn.vodadone.co.uk/shared-assets/react-is/17.0.2/umd/react-is.production.min.js",
  "styled-components": "http://cdn.vodadone.co.uk/shared-assets/styled-components/5.3.5/styled-components.min.js",
  "vue": "http://cdn.vodadone.co.uk/shared-assets/vue/3.2.25/vue.global.prod.min.js",
}
`;
